core/trino-main/src/main/java/io/trino/operator/output/UnnestingPositionsAppender.java

```diff
@@ -96,15 +116,114 @@ public long getSizeInBytes()
         }
         return IntArrayList.wrap(positionArray);
     }
+
+    private static class DictionaryBlockBuilder
+    {
+        private static final int INSTANCE_SIZE = instanceSize(DictionaryBlockBuilder.class);
+        private final int initialEntryCount;
+        private Block dictionary;
+        private int[] dictionaryIds;
+        private int positionCount;
+        private boolean closed;
+
+        public DictionaryBlockBuilder()
+        {
+            this(1024);
+        }
+
+        public DictionaryBlockBuilder(int initialEntryCount)
+        {
+            this.initialEntryCount = initialEntryCount;
+            this.dictionaryIds = new int[0];
+        }
+
+        public boolean isEmpty()
+        {
+            return positionCount == 0;
+        }
+
+        public Block build()
+        {
+            return DictionaryBlock.create(positionCount, dictionary, dictionaryIds);
+        }
+
+        public long getRetainedSizeInBytes()
+        {
+            return INSTANCE_SIZE
+                    + (long) dictionaryIds.length * Integer.BYTES
+                    + (dictionary != null ? dictionary.getRetainedSizeInBytes() : 0);
+        }
+
+        public boolean canAppend(Block dictionary)
+        {
+            return !closed && (dictionary == this.dictionary || this.dictionary == null);
+        }
+
+        public void append(IntArrayList mappedPositions, Block dictionary)
+        {
+            checkArgument(canAppend(dictionary));
+            this.dictionary = dictionary;
+            ensureCapacity(positionCount + mappedPositions.size());
+            System.arraycopy(mappedPositions.elements(), 0, dictionaryIds, positionCount, mappedPositions.size());
+            positionCount += mappedPositions.size();
+        }
+
+        public void flushDictionary(PositionsAppender delegate)
+        {
+            if (closed) {
+                return;
+            }
+            if (positionCount > 0) {
+                requireNonNull(dictionary, () -> "dictionary is null but we have pending dictionaryIds " + positionCount);
+                delegate.append(IntArrayList.wrap(dictionaryIds, positionCount), dictionary);
+            }
+
+            closed = true;
+            dictionaryIds = new int[0];
+            positionCount = 0;
+            dictionary = null;
+        }
+
+        public DictionaryBlockBuilder newBuilderLike()
+        {
+            return new DictionaryBlockBuilder(max(calculateBlockResetSize(positionCount), initialEntryCount));
+        }
+
+        private void ensureCapacity(int capacity)
+        {
+            if (dictionaryIds.length >= capacity) {
+                return;
+            }
+
+            int newSize;
+            if (dictionaryIds.length > 0) {
+                newSize = calculateNewArraySize(dictionaryIds.length);
+            }
+            else {
+                newSize = initialEntryCount;
+            }
+            newSize = Math.max(newSize, capacity);
```

Comment:
Is there any possibility that `newSize` will be lower than `capacity`?

Reply:
well yes, i.e., `capacity` can be bigger than `initialEntryCount` or bigger than `1.5 * dictionaryIds.length `(`calculateNewArraySize`)

Reply:
Why not just do `int newSize =  calculateNewArraySize(max(initialEntryCount, capacity, dictionaryIds.length))`

ofc this will work different that now. If capacity will 100, it will create an array of size 150.

Reply:
we do not want to go over `initialEntryCount` as the appender can be "pre-sized" in `#reset`